samples/ProtectedMCPClient/BasicOAuthAuthorizationProvider.cs

```diff
@@ -0,0 +1,372 @@
+                request.Headers.Authorization = new AuthenticationHeaderValue("Basic", authValue);
+            }
+            
+            var response = await _httpClient.SendAsync(request, cancellationToken);
+            if (response.IsSuccessStatusCode)
+            {
+                var json = await response.Content.ReadAsStringAsync(cancellationToken);
+                var tokenResponse = JsonSerializer.Deserialize<TokenContainer>(
+                    json, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
+                
+                if (tokenResponse != null)
+                {
+                    tokenResponse.ObtainedAt = DateTimeOffset.UtcNow;
+                    if (string.IsNullOrEmpty(tokenResponse.RefreshToken))
+                    {
+                        tokenResponse.RefreshToken = refreshToken;
+                    }
+                    
+                    return tokenResponse;
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine($"Error refreshing token: {ex.Message}");
+        }
+        
+        return null;
+    }
+
+    private async Task<TokenContainer?> InitiateAuthorizationCodeFlowAsync(
+        AuthorizationServerMetadata authServerMetadata, 
+        CancellationToken cancellationToken)
+    {
+        var codeVerifier = GenerateCodeVerifier();
+        var codeChallenge = GenerateCodeChallenge(codeVerifier);
+        
+        var authUrl = BuildAuthorizationUrl(authServerMetadata, codeChallenge);
+        
+        var authCode = await GetAuthorizationCodeAsync(authUrl, cancellationToken);
+        if (string.IsNullOrEmpty(authCode)) return null;
+        
+        return await ExchangeCodeForTokenAsync(authServerMetadata, authCode, codeVerifier, cancellationToken);
+    }
+    
+    private Uri BuildAuthorizationUrl(AuthorizationServerMetadata authServerMetadata, string codeChallenge)
+    {
+        var queryParams = HttpUtility.ParseQueryString(string.Empty);
+        queryParams["client_id"] = clientId;
+        queryParams["redirect_uri"] = _redirectUri.ToString();
+        queryParams["response_type"] = "code";
+        queryParams["code_challenge"] = codeChallenge;
+        queryParams["code_challenge_method"] = "S256";
+        
+        if (_scopes.Any())
+        {
+            queryParams["scope"] = string.Join(" ", _scopes);
+        }
+        
+        var uriBuilder = new UriBuilder(authServerMetadata.AuthorizationEndpoint);
+        uriBuilder.Query = queryParams.ToString();
+        return uriBuilder.Uri;
+    }
+    
+    private async Task<string?> GetAuthorizationCodeAsync(Uri authorizationUrl, CancellationToken cancellationToken)
+    {
+        var listenerPrefix = _redirectUri.GetLeftPart(UriPartial.Authority);
+        if (!listenerPrefix.EndsWith("/")) listenerPrefix += "/";
+        
+        using var listener = new System.Net.HttpListener();
```

Comment:
We expect clients to open ports for accepting incoming requests??

Reply:
This is for the local client sample where you have a local listener temporarily to accept the authorization code from your authorization server. This is a pattern used by MSAL, for example, for local scenarios.

Reply:
I'm not up on the terminology. Does "local client" have a special meaning in this context? If not, isn't the 99.9% case that a client is local?

Reply:
I just saw this again when resolving comments before merging this. A local client is one where there's a running process that can both open a browser and handle the redirect back all in the same process like a Windows desktop app. I'm not sure that's 99% of apps.

Another common client would be web-based ones like those running on the server in an ASP.NET Core app. If you don't have sticky-sessions or something like that, you have to use something other than the process's memory to maintain state for things like the PKCE. ASP.NET Core typically uses encrypted cookies for this purpose. I plan to demonstrate how you could implement the MCP client flow in a stateless ASP.NET Core app as a follow up.